transcom/mymove
# [MB-4529] MyMove API data load

## Description

The goal of this PR is to establish a new pattern for loading API data into the Customer app that is de-coupled from rendering UI, and can be called easily as needed throughout the flow, using Redux actions.

### New pattern is:
- Created [`rootCustomerSaga` for bootstrapping sagas that should only happen in the customer app](https://github.com/transcom/mymove/pull/4946/files#diff-c5bb6481b319d418e4307f67a9b1e8ecR10)
- Wrote [`initializeOnboarding` saga to handle hydrating onboarding user data when a customer logs in](https://github.com/transcom/mymove/pull/4946/files#diff-d41bccd15113a160657933168162f6c9R50). This should only happen once-per-page load, so [it is triggered at the highest level of the customer app (`MyMove/index componentDidMount`)](https://github.com/transcom/mymove/pull/4946/files#diff-6eb50225743dcbdf87fad2a63022f3a4R66), and it handles the following:
  - Get logged in user
  - If user doesn't have a service member, create one
  - If user has moves, load MTO shipments also
  - Populate user data & MTO shipments into Entities
- All API calls should be defined in `src/services` and use the [new `makeSwaggerRequest` function](https://github.com/transcom/mymove/pull/4946/files#diff-eadb579699721add54d5fa7d9853ef3cR78)
  - Existing API calls should be relocated from entities files as needed
  - `makeSwaggerRequest` accepts a `normalize` option (defaults to true) so that API calls can be used to populate both entities _and_ ducks state for backwards compatibility, and to safely proceed with Redux refactor work.

### Proof-of-concepts:
- Replaced `loadMTOShipments` in `HHGShipmentSetup componentDidMount` with new `fetchCustomerData` action, to demonstrate how to refresh _all_ customer data into entities
- Added `initializeOnboarding` action to `MyMove/index componentDidMount` to kick off new onboarding API flow 
- Removed `showLoggedInUser` and `loadMTOShipments` in `MyMove/Home/index componentDidMount, componentDidUpdate` since those calls will happen as part of `initializeOnboarding`
- Removed `createServiceMember` in `MyMove/Home/index componentDidUpdate` since that will happen as part of `initializeOnboarding`
- Added `updateServiceMember` saga to populate service member entities throughout onboarding steps in order to pass tests (this will get cleaned up when service member ducks are refactored)

### Next steps:
- The customer app is calling `getLoggedInUser` twice (once in `initializeOnboarding` and once in `loadUser`) -- clean this up so only one call is made, in a way that is compatible with the office app.
- Identify any other API calls that should be triggered when a customer user logs in (in addition to MTO shipments)
- Identify other places where re-fetch API calls may not be needed anymore and can be deleted, or replaced with `fetchCustomerData` (or something smaller)
- **Continue to replace API calls with sagas so they can update data in ducks & entities simultaneously for backwards compatibility**
  - Once this is completed, we can continue to remove ducks code with more reliability 🎉 

## Reviewer Notes

Test cases:
- Log in as existing MilMove customer & verify your data loads and you are directed to the right place
- Log in as new MilMove customer & verify your service member is created and you are directed to the right place

## Setup

```
make server_run
make client_run
```

## Code Review Verification Steps

 [Querying the Database Safely](https://github.com/transcom/mymove/wiki/Backend-Programming-Guide#querying-the-database-safely)
 have been satisfied.
* Any new migrations/schema changes:

## References

* [Jira story](https://dp3.atlassian.net/browse/MB-4529) for this change

## Screenshots

Here is a fun diagram I made that illustrates the saga flows: 
![image](https://user-images.githubusercontent.com/2723066/95790638-647d4300-0ca5-11eb-81cf-17b2fb774821.png)
